assumeutxo, rpc: Improve EOF error when reading snapshot metadata in loadtxoutset

pablomartin4btc opened the pull request:
Improve error messaging in `loadtxoutset` to provide a more user-friendly error message when the utxo snapshot file size is insufficient to contain the necessary metadata structure (40 bytes).

- Error message output before the change (current master):
```
./src/bitcoin-cli -datadir=${AU_DATADIR} loadtxoutset ${AU_DATADIR}/utxo-111.dat
error code: -1
error message:
AutoFile::read: end of file: iostream error
```

- Error message output after the change:
```
./src/bitcoin-cli -datadir=${AU_DATADIR} loadtxoutset ${AU_DATADIR}/utxo-111.dat
error code: -32603
error message:
Unable to load UTXO snapshot, couldn't read snapshot metadata from file /tmp/.test_utxo_3/utxo-111.dat.
The file may be corrupted or it was crafted in an incompatible format.
```


Original strategy was to check the size of the file before de-serialisation and the current strategy was presented at that time as an alternative which has been preferred by reviewers and it looks simpler since `streams` source code files are no longer touched.

This PR needs to be backported to 26.x

pablomartin4btc committed:
assumeutxo: improve EOF error when reading snapshot

In the RPC command `loadtxoutset`, catch possible errors while
reading snapshot metadata, most commonly due to the structure size
being shorter than the expected one (40 bytes) or the snapshot file
may be corrupted or crafted in an incompatible format.

kevkevinpal commented on src/rpc/blockchain.cpp:
```diff
@@ -2749,6 +2749,11 @@ static RPCHelpMan loadtxoutset()
     SnapshotMetadata metadata;
+    // Check if the file size is at least as large as the expected size of SnapshotMetadata
+    if (afile.size() < sizeof(SnapshotMetadata)) {
+        throw JSONRPCError(RPC_INTERNAL_ERROR, strprintf("Unable to load UTXO snapshot, "
```
nit: should we specify that the file size is too small? A user might think their file is too large with this error message

kevkevinpal replied:
just saw your alternative, that is preferred for me

pablomartin4btc replied:
My intention with the code change alternative was to avoid touching both stream files, header and .cpp, but sure, I could take the error message from there and use it here.
Regarding the error message output, I'm open to suggestions there, I could specify that the file's too small or the exact minimum amount of bytes that are required, or even ask to check the documentation, which will need to be updated.

pablomartin4btc commented:
Rebased + fix correct usage of `utf8string()` instead of `u8string()`.

furszy reviewed (commented):
Would be good to add test coverage for this error.

Also, the "file size being too small" error message isn't really descriptive and it does not provide further guidance to solve the problem. The file could be corrupt or could had been crafted erroneously.
I would go simpler for now and change "file size being too small" for "Unable to load UTXO snapshot, the file %s cannot be parsed."

I see your alternative solution better. More comprehensive.

pablomartin4btc committed:
test, assumeutxo: add coverage for loadtxoutset EOF error

Add a test to cover scenarios where an EOF error is raised during
the reading of invalid snapshot metadata (40 bytes file size min)
or a corrupted file with an invalid format.

pablomartin4btc commented:
Updates:
- Switch the logic strategy to the previously presented alternative due to preference of reviewers @kevkevinpal & @furszy (thanks!);
- Added test coverage

furszy commented on src/rpc/blockchain.cpp:
```diff
@@ -2746,7 +2746,15 @@ static RPCHelpMan loadtxoutset()
+        // Handle any exception, which includes reaching the end of the file e.g. file size < sizeOf(metadata)
+        throw JSONRPCError(RPC_INTERNAL_ERROR, strprintf("Unable to load UTXO snapshot, "
+            "couldn't read snapshot metadata from file %s.\nThis error is usually "
+            "due to the file size being too small.\n%s", path.utf8string(), e.what()));
```
I don't to see how this part of the message would be helpful to users. "file size being too small" does not tell me anything. Would be better to say that the file is corrupt and/or was crafted in an incompatible format. Also, no need to add the internal exception message to the user facing error.

pablomartin4btc replied:
Makes sense, thanks!

pablomartin4btc commented:
Addressed @furszy's [feedback](https://github.com/bitcoin/bitcoin/pull/28670#discussion_r1456421159).

hebasto commented on src/rpc/blockchain.cpp:
```diff
@@ -2746,7 +2746,15 @@ static RPCHelpMan loadtxoutset()
+    try {
+        // Read data from the file
+        afile >> metadata;
+    } catch (const std::exception& e) {
```
To fix the MSVC CI job, suggesting:

```suggestion
    } catch (const std::exception&) {
```

pablomartin4btc replied:
done, thanks!

pablomartin4btc commented:
Addressed @hebasto's [suggestion](https://github.com/bitcoin/bitcoin/pull/28670#discussion_r1457712659) in order to fix the MSVC CI job.

hernanmarino reviewed (approved):
Approach ACK

fjahr commented:
Code review ACK 53fbf19e95a3d5083eaf033378bb8d0e87bdc6b2

pablomartin4btc commented:
Updates:
- I think [lint CI failure](https://github.com/bitcoin/bitcoin/pull/28670/checks?check_run_id=20626357772) is unrelated with the code change, it was running successful before started failing with no code change, run `./test/lint/lint-python.py` locally with no failure.

furszy reviewed (commented):
The CI failure is due a merge conflict. Need to rebase the PR. Some of the code you are using here is not in master anymore.

pablomartin4btc commented:
Rebased.

fjahr commented on test/functional/feature_assumeutxo.py:
```diff
@@ -114,6 +114,13 @@ def expected_error(log_msg="", rpc_details=""):
                 f.write(valid_snapshot_contents[(32 + 8 + offset + len(content)):])
             expected_error(log_msg=f"[snapshot] bad snapshot content hash: expected a4bf3407ccb2cc0145c49ebba8fa91199f8a3903daf0883875941497d2493c27, got {wrong_hash}")
 
+        self.log.info("  - snapshot file with invalid size, too small")
```
nit: I think this info log could be improved if you retouch. Testing the file size is not really the goal, it's about hitting the metadata parsing error that was added, right? So I would suggest to call it something like "Inablity to parse metadata fails early" or so.

pablomartin4btc replied:
Thanks! I agree, I'll update it if I have to retouch it and if not, could be done in a follow-up.

pablomartin4btc replied:
done!

fjahr reviewed (commented):
re-ACK 521de52c751a4539333bb2f69a07c0f1b4f496de

furszy commented on src/rpc/blockchain.cpp:
```diff
@@ -2771,7 +2771,15 @@ static RPCHelpMan loadtxoutset()
+        afile >> metadata;
+    } catch (const std::exception&) {
+        // Handle any exception, which includes reaching the end of the file e.g. file size < sizeOf(metadata)
+        throw JSONRPCError(RPC_INTERNAL_ERROR, strprintf("Unable to load UTXO snapshot, "
```
Shouldn't use `RPC_INTERNAL_ERROR`. As per it description: "this error should only be used for genuine errors in bitcoind (for example datadir corruption)"

Maybe use `RPC_DESERIALIZATION_ERROR` or `RPC_MISC_ERROR.

pablomartin4btc replied:
It makes sense, I'll fix it, thanks!

pablomartin4btc replied:
done!

furszy reviewed (commented):
utACK 521de52c751

pablomartin4btc commented:
Updates:
- Addressed [feedback](https://github.com/bitcoin/bitcoin/pull/28670#discussion_r1566860922) from @fjahr by amending the logging info description.
- Addressed [feedback](https://github.com/bitcoin/bitcoin/pull/28670#discussion_r1597147718) from @furszy by using a better RPC error code (`RPC_DESERIALIZATION_ERROR` instead of `RPC_INTERNAL_ERROR`).

furszy commented on src/rpc/blockchain.cpp:
```diff
@@ -2778,7 +2778,15 @@ static RPCHelpMan loadtxoutset()
+        // Handle any exception, which includes reaching the end of the file e.g. file size < sizeOf(metadata)
+        throw JSONRPCError(RPC_DESERIALIZATION_ERROR, strprintf("Unable to load UTXO snapshot, "
+            "couldn't read snapshot metadata from file %s.\nThe file may be corrupted "
+            "or it was crafted in an incompatible format.", path.utf8string()));
```
nit: should use `fs::PathToString(path)`

furszy reviewed (commented):
utACK 4a2df05e

pablomartin4btc commented:
> Hm, unfortunately, I think this PR was made redundant by #29612 recently. After the latest feedback there which I addressed just 3 weeks ago, we are enhancing the assumeutxo metadata there significantly and this means all of the lines changed here are touched there as well and I don't see much of an overlap either that would let me use the code here.

No prob, thanks for letting me know, I'll review that PR.

pablomartin4btc commented:
> Hm, unfortunately, I think this PR was made redundant by #29612 recently. After the latest feedback there which I addressed just 3 weeks ago, we are enhancing the assumeutxo metadata there significantly and this means all of the lines changed here are touched there as well and I don't see much of an overlap either that would let me use the code here.

No prob, thanks for letting me know, I'll review that PR.
